tests/basic.json

```diff
@@ -400,6 +400,15 @@
         1
       ]
     },
+    {
+      "name": "descendant segment, wildcard selector, nested arrays",
+      "selector" : "$..[*]",
+      "document" : [[[1]],[2]],
+      "results": [
+        [[[1]],[2],[1],1,2],
+        [[[1]],[2],[1],2,1]
```

Comment:
I expect we'd want all of the orderings here.

These should be valid as well, right?

```
[[[1]],[1],1,[2],2]
[[[1]],[1],[2],1,2]
[[[1]],[1],[2],2,1]
[[[1]],[2],2,[1],1]
```

I think the only requirements are:

- `[[1]]` comes before `[2]` (array order) and `[1]` (descendant)
- `[2]` comes before `2` (descendant)
- `[1]` comes before `[1]` (descendant)

Reply:
I wonder if there's a way we can express the constraints themselves rather than listing all of the possibilities.

Reply:
> These should be valid as well, right?

Although valid permutations, applying the wildcard selector to `1` and `2` will never produce a node, so their position has no impact on the resulting node list.

Reply:
> applying the wildcard selector to 1 and 2 will never produce a node

They don't have to produce a node for these orderings.  I'm not sure what point you're making.

Reply:
I might have misunderstood what you were saying 😞

Reply:
> I expect we'd want all of the orderings here.
> 
> These should be valid as well, right?
> 
> ```
> [[[1]],[1],1,[2],2]
> [[[1]],[1],[2],1,2]
> [[[1]],[1],[2],2,1]
> [[[1]],[2],2,[1],1]
> ```

No, these are the result of `..` before `[*]` is applied. See [section 2.5.2.2](https://www.rfc-editor.org/rfc/rfc9535#name-semantics-9) of the RFC. Actually, I made the same mistake the other day, see "Confession" in my recent [blog post](https://underlap.org/testing-non-determinism).

> 
> I think the only constraints are:
> 
>     * `[[1]]` comes before `[2]` (array order) and `[1]` (descendant)
> 
>     * `[2]` comes before `2` (descendant)
> 
>     * `[1]` comes before `1` (descendant)

Those are constraints on the traversal order. The RFC requires that `[*]` is applied to the input node and its descendants.

Reply:
> I wonder if there's a way we can express the constraints themselves rather than listing all of the possibilities.

@gregsdennis I've given that a lot of thought. The best description of the constraints that I know of is in the RFC, but that's not amenable to coding in a testcase. The difficulty is that a mixture of deterministically and non-deterministically ordered subsections are flattened into the resultant nodelist and it's very hard to keep track of what's going on. If someone can think of a way of encoding this rather than listing all the possibilities, that would be very interesting. For the moment, I think generating the list of possibilities is the most reliable approach.

Reply:
> these are the result of `..` before `[*]` is applied.

This doesn't make sense `..` on its own doesn't do anything.  It merely informs the `[*]` that it should recurse.

Reply:
If you're confident that these are the only two possibilities, I'm happy to go with it.  I still don't understand why the others I listed aren't valid outputs, though.